### PR TITLE
[Feature] Throttle Policy Reset Support for APIM

### PR DESCRIPTION
**Pull request by Chanuka-ChandraYapa**
Related Issue: https://github.com/wso2/api-manager/issues/2455

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.throttling.siddhi.extension/src/test/java/org/wso2/carbon/apimgt/throttling/siddhi/extension/CountAttributeAggregatorWithResetTestCase.java [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+import org.junit.Test;
+import org.wso2.siddhi.core.ExecutionPlanRuntime;
+import org.wso2.siddhi.core.SiddhiManager;
+import org.wso2.siddhi.core.event.Event;
+import org.wso2.siddhi.core.query.output.callback.QueryCallback;
+import org.wso2.siddhi.core.stream.input.InputHandler;
+import org.wso2.siddhi.core.util.EventPrinter;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
```
Comment (rmsamitha) [inline components/apimgt/org.wso2.carbon.apimgt.throttling.siddhi.extension/src/test/java/org/wso2/carbon/apimgt/throttling/siddhi/extension/CountAttributeAggregatorWithResetTestCase.java:33]: Java doc comments missing for the class

File: components/apimgt/org.wso2.carbon.apimgt.throttling.siddhi.extension/src/test/java/org/wso2/carbon/apimgt/throttling/siddhi/extension/CountAttributeAggregatorWithResetTestCase.java [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+public class CountAttributeAggregatorWithResetTestCase {
+    private static final Log log = LogFactory.getLog(CountAttributeAggregatorWithResetTestCase.class);
+    private AtomicInteger atomicEventCount;
+
+    @Before
+    public void init() {
+        atomicEventCount = new AtomicInteger(0);
+    }
+
```
Comment (rmsamitha) [inline components/apimgt/org.wso2.carbon.apimgt.throttling.siddhi.extension/src/test/java/org/wso2/carbon/apimgt/throttling/siddhi/extension/CountAttributeAggregatorWithResetTestCase.java:42]: Java doc comments missing. Mention what is tested by this test

Comment (rmsamitha) [reply components/apimgt/org.wso2.carbon.apimgt.throttling.siddhi.extension/src/test/java/org/wso2/carbon/apimgt/throttling/siddhi/extension/CountAttributeAggregatorWithResetTestCase.java:42]: Do for other test methods as well

File: components/apimgt/org.wso2.carbon.apimgt.throttling.siddhi.extension/src/test/java/org/wso2/carbon/apimgt/throttling/siddhi/extension/CountAttributeAggregatorWithResetTestCase.java [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+            @Override
+            public void receive(long timeStamp, Event[] inEvents, Event[] removeEvents) {
+                atomicEventCount.set(0);
+                EventPrinter.print(timeStamp, inEvents, removeEvents);
+                if (inEvents != null) {
+                    atomicEventCount.addAndGet(inEvents.length);
+                    if (atomicEventCount.get() == 1) {
+                        junit.framework.Assert.assertEquals("APIM", inEvents[0].getData(0));
+                        junit.framework.Assert.assertEquals(1L, inEvents[0].getData(1));
+
```
Comment (rmsamitha) [inline components/apimgt/org.wso2.carbon.apimgt.throttling.siddhi.extension/src/test/java/org/wso2/carbon/apimgt/throttling/siddhi/extension/CountAttributeAggregatorWithResetTestCase.java:173]: Code not formatted. Empty lines are there

File: components/apimgt/org.wso2.carbon.apimgt.throttling.siddhi.extension/src/test/java/org/wso2/carbon/apimgt/throttling/siddhi/extension/SumAttributeAggregatorWithResetTestCase.java [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+import org.junit.Test;
+import org.wso2.siddhi.core.ExecutionPlanRuntime;
+import org.wso2.siddhi.core.SiddhiManager;
+import org.wso2.siddhi.core.event.Event;
+import org.wso2.siddhi.core.query.output.callback.QueryCallback;
+import org.wso2.siddhi.core.stream.input.InputHandler;
+import org.wso2.siddhi.core.util.EventPrinter;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
```
Comment (rmsamitha) [inline components/apimgt/org.wso2.carbon.apimgt.throttling.siddhi.extension/src/test/java/org/wso2/carbon/apimgt/throttling/siddhi/extension/SumAttributeAggregatorWithResetTestCase.java:33]: Class level doc comment missing

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java [hunk 5]
```diff
@@ -4344,6 +4345,36 @@ public Tier getThrottlePolicyByName(String policyId, int policyType, String orga
...
+        String groupId = application.getGroupId();
+        int appId = application.getId();
+        String appTier = application.getTier();
+        // Check whether the application is accessible to the logged-in user
+        String loggedInUser = (userNameWithoutChange != null) ? userNameWithoutChange : username;
+        if (!validateApplication(loggedInUser, appId, groupId)) {
+            log.error("Application " + applicationId + " is not accessible to user " + loggedInUser);
+            throw new APIMgtAuthorizationFailedException(
+                    "Application is not accessible to user  " + loggedInUser);
+
```
[the remark this answers is not in the document]
Comment (Chanuka-ChandraYapa) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/12443/commits/a583c6ae8e5be9a7a7879183801ad442b86123e4#diff-eea96b2f615001930c3a7cf2ec3802f9dc86ac99cabcd2ae10716b934c76c3a2

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApplicationsApiServiceImpl.java [hunk 6]
```diff
@@ -486,6 +488,39 @@ public Response applicationsApplicationIdPut(String applicationId, ApplicationDT
         return null;
     }
 
+    /**
+     * Reset Application Level Throttle Policy
+     *
+     * @param applicationId application Identifier
+     * @param applicationThrottleResetDTO request DTO containing the username
+     * @return 200 if successful
```
[the remark this answers is not in the document]
Comment (Chanuka-ChandraYapa) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApplicationsApiServiceImpl.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/12443/commits/a583c6ae8e5be9a7a7879183801ad442b86123e4#diff-12f710906e8ac379963b9527b560b589024a79f07c7395c425241e517048a0b0

File: components/apimgt/org.wso2.carbon.apimgt.throttle.policy.deployer/src/main/java/org/wso2/carbon/apimgt/throttle/policy/deployer/utils/ThrottlePolicyJMSMessageListener.java [hunk 7]
```diff
@@ -209,6 +218,48 @@ private void handleNotificationMessage(String eventType, String encodedEvent) {
...
+                    applicationLevelTier, null, null, null, null, null, null, authorizedUser, null, null,
+                    subscriberTenantDomain, null, applicationId, null, jsonObMap.toString() };
+            EventPublisherEvent resetEvent = new EventPublisherEvent(streamID,
+                    System.currentTimeMillis(), objects, loggingEvent);
+            resetEvent.setOrgId(event.getTenantDomain());
+            if (log.isDebugEnabled()) {
+                log.debug(
+                        "Publishing Application Throttle Reset data to traffic-manager for the application with ID: "
+                                + applicationId + " started" + " at " + new SimpleDateFormat(
+                                "[yyyy.MM.dd HH:mm:ss,SSS zzz]").format(new Date()));
```
[the remark this answers is not in the document]
Comment (Chanuka-ChandraYapa) [reply components/apimgt/org.wso2.carbon.apimgt.throttle.policy.deployer/src/main/java/org/wso2/carbon/apimgt/throttle/policy/deployer/utils/ThrottlePolicyJMSMessageListener.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/12443/commits/a583c6ae8e5be9a7a7879183801ad442b86123e4#diff-297698e5fec79a02c2164251f303435e44417e5d79d7f59bf3fd8182d653e405

File: components/apimgt/org.wso2.carbon.apimgt.throttle.policy.deployer/src/main/java/org/wso2/carbon/apimgt/throttle/policy/deployer/utils/ThrottlePolicyJMSMessageListener.java [hunk 8]
```diff
@@ -209,6 +218,48 @@ private void handleNotificationMessage(String eventType, String encodedEvent) {
...
+            }
+            APIUtil.publishEvent(EventPublisherType.NOTIFICATION, resetEvent, loggingEvent);
+            if (log.isDebugEnabled()) {
```
[the remark this answers is not in the document]
Comment (Chanuka-ChandraYapa) [reply components/apimgt/org.wso2.carbon.apimgt.throttle.policy.deployer/src/main/java/org/wso2/carbon/apimgt/throttle/policy/deployer/utils/ThrottlePolicyJMSMessageListener.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/12443/commits/a583c6ae8e5be9a7a7879183801ad442b86123e4#diff-297698e5fec79a02c2164251f303435e44417e5d79d7f59bf3fd8182d653e405

File: components/apimgt/org.wso2.carbon.apimgt.throttling.siddhi.extension/src/test/java/org/wso2/carbon/apimgt/throttling/siddhi/extension/CountAttributeAggregatorWithResetTestCase.java (hunk 1 above)
Comment (Chanuka-ChandraYapa) [reply components/apimgt/org.wso2.carbon.apimgt.throttling.siddhi.extension/src/test/java/org/wso2/carbon/apimgt/throttling/siddhi/extension/CountAttributeAggregatorWithResetTestCase.java:33]: Fixed with https://github.com/wso2/carbon-apimgt/pull/12443/commits/a583c6ae8e5be9a7a7879183801ad442b86123e4#diff-a92a928154247caabe1158c2ac06adfeab5798b6f9e50538cb203bffabb50a70

File: components/apimgt/org.wso2.carbon.apimgt.throttling.siddhi.extension/src/test/java/org/wso2/carbon/apimgt/throttling/siddhi/extension/CountAttributeAggregatorWithResetTestCase.java (hunk 3 above)
Comment (Chanuka-ChandraYapa) [reply components/apimgt/org.wso2.carbon.apimgt.throttling.siddhi.extension/src/test/java/org/wso2/carbon/apimgt/throttling/siddhi/extension/CountAttributeAggregatorWithResetTestCase.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/12443/commits/a583c6ae8e5be9a7a7879183801ad442b86123e4#diff-a92a928154247caabe1158c2ac06adfeab5798b6f9e50538cb203bffabb50a70

File: components/apimgt/org.wso2.carbon.apimgt.throttling.siddhi.extension/src/test/java/org/wso2/carbon/apimgt/throttling/siddhi/extension/SumAttributeAggregatorWithResetTestCase.java (hunk 4 above)
Comment (Chanuka-ChandraYapa) [reply components/apimgt/org.wso2.carbon.apimgt.throttling.siddhi.extension/src/test/java/org/wso2/carbon/apimgt/throttling/siddhi/extension/SumAttributeAggregatorWithResetTestCase.java:33]: Fixed with https://github.com/wso2/carbon-apimgt/pull/12443/commits/a583c6ae8e5be9a7a7879183801ad442b86123e4#diff-a92a928154247caabe1158c2ac06adfeab5798b6f9e50538cb203bffabb50a70

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.common/src/main/resources/devportal-api.yaml [hunk 9]
```diff
@@ -3143,6 +3143,55 @@ paths:
         - lang: Curl
           source: curl -k "https://localhost:9443/api/am/devportal/v3/throttling-policies/application/10PerMin"
```
[the remark this answers is not in the document]
Comment (Chanuka-ChandraYapa) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.common/src/main/resources/devportal-api.yaml:3145]: common module update their .yaml files automatically when building

File: components/apimgt/org.wso2.carbon.apimgt.throttling.siddhi.extension/src/test/java/org/wso2/carbon/apimgt/throttling/siddhi/extension/CountAttributeAggregatorWithResetTestCase.java (hunk 2 above)
Comment (Chanuka-ChandraYapa) [reply components/apimgt/org.wso2.carbon.apimgt.throttling.siddhi.extension/src/test/java/org/wso2/carbon/apimgt/throttling/siddhi/extension/CountAttributeAggregatorWithResetTestCase.java:45]: Added meaningful method names with https://github.com/wso2/carbon-apimgt/pull/12443/commits/a583c6ae8e5be9a7a7879183801ad442b86123e4#diff-a92a928154247caabe1158c2ac06adfeab5798b6f9e50538cb203bffabb50a70

**Review by rmsamitha — APPROVED** [review]

**Pull request merged**